### PR TITLE
fix(Core/Chat): Emotes (/e Text) in game

### PR DESCRIPTION
**Pull request by poszer**
Co-Authored-By: Lee <talamortis@users.noreply.github.com>

##### CHANGES PROPOSED:

-  This will fix issue with gamebreaking issue where players can speak with opposite faction using emotes. For example : /e Text
Thanks to @talamortis
-  

###### ISSUES ADDRESSED:

Closes https://github.com/azerothcore/azerothcore-wotlk/issues/2099

##### TESTS PERFORMED:

Yes, tested in game. Looks fine now, The players can see the custom emotes from their factions, but can't see the custom emotes from opposite factions.
The normal (default) emotes like /laugh /smile /welcome /cheer are visible for everyone.

##### HOW TO TEST THE CHANGES:

1. Try to speak using emote commands between Alliance & Horde character
2. Example:  /e Hello how are you

##### KNOWN ISSUES AND TODO LIST:

- [ ]
- [ ] 

##### Target branch(es):

Master

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -20911,16 +20911,28 @@ void Player::Yell(const std::string& text, const uint32 language)
     SendMessageToSetInRange(&data, sWorld->getFloatConfig(CONFIG_LISTEN_RANGE_YELL), true);
 }
 
-void Player::TextEmote(const std::string& text)
+void Player::TextEmote(const std::string& text, uint32 type)
 {
     std::string _text(text);
-    sScriptMgr->OnPlayerChat(this, CHAT_MSG_EMOTE, LANG_UNIVERSAL, _text);
+    sScriptMgr->OnPlayerChat(this, CHAT_MSG_EMOTE, LANG_UNIVERSAL, _text); // It was like this b
```
Comment (talamortis) [inline src/server/game/Entities/Player/Player.cpp]: // It was like this b needs removing :P

Comment (poszer) [reply src/server/game/Entities/Player/Player.cpp]: Yeah, done. Thanks :smile:

**Comment by poszer** [comment]
Issue with Eluna >.< For users who don't have Eluna module, this can be tested without problem even if travis fail

**Comment by poszer** [comment]
> This does sufficiently prevent emotes from passing between factions, however, what should be displayed if a player types a custom emote is '%playername% makes some strange gestures.'
> 
> Additionally, items that create an emote, such as placing a Fish Feast (Item ID: 45279) do not display to the opposite faction, but these emotes should be shown to both factions.
> 
> Additionally, emotes like flipping a Dalaran coin or placing a Fish Feast are not properly adding a Player's name to the emote. It just says 'prepares a Fish Feast!', instead of '%playername% prepares a Fish Feast!'.
> Annotation 2019-07-21 095320

Good, will be improved. Thanks for testing

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -20911,16 +20911,28 @@ void Player::Yell(const std::string& text, const uint32 language)
...
 #endif
+
     WorldPacket data;
-    ChatHandler::BuildChatPacket(data, CHAT_MSG_EMOTE, LANG_UNIVERSAL, this, this, _text);
+    switch (type)
+    {
+    case 1: // This will be for normal emotes
+        ChatHandler::BuildChatPacket(data, CHAT_MSG_EMOTE, LANG_UNIVERSAL, this, this, _text);
+        break;
+    case 2: // This wll be for custom.
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.cpp]: typo in `wll`

File: src/server/scripts/Spells/spell_item.cpp [hunk 3]
```diff
@@ -1059,11 +1059,11 @@ class spell_item_oracle_ablutions : public SpellScriptLoader
...
                 {
                     std::string name = player->GetName();
-                    player->TextEmote("casually flips his Titanium Seal of Dalaran");
+                    player->TextEmote("casually flips his Titanium Seal of Dalaran", 1);
                     if (urand(0,1))
-                        player->TextEmote("finds the coin face down for tails!");
+                        player->TextEmote("finds the coin face down for tails!", 1);
                     else
-                        player->TextEmote("catches the coin heads up!");
+                        player->TextEmote("catches the coin heads up!", 1);
```
Comment (FrancescoBorzi) [inline src/server/scripts/Spells/spell_item.cpp]: I would prefer wrapping those two values `1` and `2` into enums and use them accordingly, so it makes the code more readable. For example something like:

```
EMOTE_TYPE_UNIVERSAL = 1
EMOTE_TYPE_FACTION_ONLY = 2
```

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -20911,16 +20911,28 @@ void Player::Yell(const std::string& text, const uint32 language)
...
+        if (this->GetTeamId() == TEAM_ALLIANCE)
+            ChatHandler::BuildChatPacket(data, CHAT_MSG_TEXT_EMOTE, LANG_COMMON, this, this, _text);
+        else
+            ChatHandler::BuildChatPacket(data, CHAT_MSG_TEXT_EMOTE, LANG_ORCISH, this, this, _text);
+        break;
+    }
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Player/Player.cpp]: cases inside switch are not indented

**Comment by poszer** [comment]
Ok, done. Ready for test @heyaapl

**Comment by andoys** [comment]
Tested and confirmed working! Finally won't abuse this command anymore to talk with each other

**Comment by talamortis** [comment]
Going to rework this with @poszer  tomorrow (Wednesday 24th 2019) to support travis & eluna aswell.

**Comment by BarbzYHOOL** [comment]
was this reworked or not?

**Comment by Ramys** [comment]
Does it work?

**Comment by Stoabrogga** [comment]
@BarbzYHOOL 
I think in order to get this all working you need the following steps:
- Create a new PR containing all changes of this PR, as the branch here was deleted (easy)
- Disable the Eluna module for Travis in this new PR and merge it (Eluna will not work now, but that's only temporarily)
- Create a new PR for LuaEngine PlayerMethods, because the TextEmote now uses 1 more parameter for AC:
  https://github.com/ElunaLuaEngine/Eluna/blob/master/PlayerMethods.h#L3067
- After the PR for LuaEngine is merged, create a new PR for the Eluna module to use the latest LuaEngine and merge it
- Create a new PR to enable Eluna for Travis once more and merge it (Eluna will now work again)

I can take a look at this if you want.

**Comment by Stoabrogga** [comment]
On second thought, there could be an easier method to solving this... let me think about this.

**Pull request closed without merging**